### PR TITLE
Compact fetch schema on read is implemented. [Api 1327]

### PR DESCRIPTION
**Pull request by OzanCansel**
- Map write/read tests are added.
- Compact schema fetch tests are added.
- Compact mix tests are added.

### CODE REVIEW

File: hazelcast/include/hazelcast/client/protocol/ClientMessage.h [hunk 1]
```diff
@@ -1207,6 +1253,12 @@ class HAZELCAST_API ClientMessage
         header->flags = is_final ? IS_FINAL_FLAG : DEFAULT_FLAGS;
         std::memcpy(
           fp + SIZE_OF_FRAME_LENGTH_AND_FLAGS, &bytes[0], bytes.size());
+
+        copy(begin(value.schemas_will_be_replicated()),
+             end(value.schemas_will_be_replicated()),
+             back_inserter(schemas_will_be_replicated_));
+
+        contains_serialized_data_in_request_ = true;
```
Comment (akeles85) [inline hazelcast/include/hazelcast/client/protocol/ClientMessage.h:1261]: I could not find the usage of this variable. Is it for debug purpose?

File: hazelcast/include/hazelcast/client/serialization/pimpl/compact/compact_impl.h [hunk 2]
```diff
@@ -574,18 +570,59 @@ struct schema_of
...
 T inline compact_stream_serializer::read(object_data_input& in)
 {
     int64_t schema_id = in.read<int64_t>();
-    const auto& local_schema = schema_of<T>::schema_v;
-    // optimization to avoid hitting shared map in the schema_service,
-    // in the case incoming data's schema is same as the local schema
-    if (schema_id == local_schema.schema_id()) {
-        compact_reader reader = create_compact_reader(*this, in, local_schema);
+
+    const boost::optional<schema>& local_schema = class_to_schema<T>::get();
```
Comment (akeles85) [inline hazelcast/include/hazelcast/client/serialization/pimpl/compact/compact_impl.h:619]: same issue in PR #1039. Just want to mention again for merging.

File: hazelcast/include/hazelcast/client/spi/impl/ClientInvocation.h [hunk 3]
```diff
@@ -32,6 +32,10 @@
 #include "hazelcast/client/spi/EventHandler.h"
 #include "hazelcast/client/protocol/ClientMessage.h"
 
+#include "hazelcast/client/serialization/pimpl/compact/default_schema_service.h"
+
+#include "hazelcast/client/serialization/pimpl/compact/default_schema_service.h"
```
Comment (akeles85) [inline hazelcast/include/hazelcast/client/spi/impl/ClientInvocation.h:37]: typo

File: hazelcast/src/hazelcast/client/network.cpp [hunk 4]
```diff
@@ -647,6 +683,108 @@ ClientConnectionManagerImpl::check_client_active()
...
+                    "Failure during sending state to the cluster. %1%"
+                ) % e.what()
+            ).str()
+        );
+
+        {
+            std::lock_guard<std::recursive_mutex> guard(client_state_mutex_);
+
+            if (cluster_id_ == target_cluster_id){
+                if (logger_.enabled(hazelcast::logger::level::fine)){
```
Comment (akeles85) [inline hazelcast/src/hazelcast/client/network.cpp:764]: is there a specific reason of not using HZ_LOG?

**Comment by akeles85** [comment]
As I see, PR contains a rebase of 1039. I did not give comments for that rebase as I already gave them before in PR #1039

File: hazelcast/include/hazelcast/client/serialization/pimpl/compact/compact_impl.h (hunk 2 above)
Comment (OzanCansel) [reply hazelcast/include/hazelcast/client/serialization/pimpl/compact/compact_impl.h]: Fixed.

File: hazelcast/include/hazelcast/client/spi/impl/ClientInvocation.h (hunk 3 above)
Comment (OzanCansel) [reply hazelcast/include/hazelcast/client/spi/impl/ClientInvocation.h]: Fixed.

File: hazelcast/src/hazelcast/client/network.cpp [hunk 5]
```diff
@@ -561,7 +580,17 @@ ClientConnectionManagerImpl::connect_to_cluster()
     if (async_start_) {
         submit_connect_to_cluster_task();
     } else {
-        do_connect_to_cluster();
+        if (do_connect_to_cluster()){
+            return;
+        }
+
+        {
+            std::lock_guard<std::recursive_mutex> guard {client_state_mutex_};
```
[the remark this answers is not in the document]
Comment (OzanCansel) [reply hazelcast/src/hazelcast/client/network.cpp]: Fixed.

File: hazelcast/include/hazelcast/client/protocol/ClientMessage.h [hunk 6]
```diff
@@ -917,6 +917,44 @@ class HAZELCAST_API ClientMessage
...
+        using namespace serialization;
+        using namespace serialization::pimpl;
+
+        // skip begin frame
+        skip_frame();
+
+        auto type_name = get<std::string>();
+
+        std::unordered_map<std::string, field_descriptor> fields;
+        {
```
Comment (ihsandemir) [inline hazelcast/include/hazelcast/client/protocol/ClientMessage.h:934]: Better to do like this:
`auto fields_list = get<std::vector<field_descriptor>>();`

and have schema constructor works with list as well.

**Review by ihsandemir — CHANGES_REQUESTED** [review]
to be continued.

File: hazelcast/src/hazelcast/client/compact.cpp [hunk 7]
```diff
@@ -1887,12 +1887,36 @@ default_schema_service::get(int64_t schemaId)
...
+    response.skip_frame();
+    auto sch = response.get_nullable<schema>();
+
+    if (!sch) {
         throw exception::illegal_state{ "default_schema_service::get",
                                         "Schema doesn't exist for this type" };
     }
 
-    return *ptr;
+    return *sch;
```
Comment (ihsandemir) [inline hazelcast/src/hazelcast/client/compact.cpp:1919]: does this do move or copy?

File: hazelcast/test/src/compact/compact_schema_replication_on_cluster_restart_test.h [hunk 8]
```diff
@@ -54,6 +79,20 @@ TEST_F(CompactSchemaReplicationOnClusterRestart, on_cluster_restart)
...
+{
+    auto condition = [this]() { return number_of_schemas_on_member() == 0; };
+
+    ASSERT_TRUE_EVENTUALLY(condition());
+
+    member_.shutdown();
+
+    HazelcastServer another_member{ factory_ };
+
+    ASSERT_TRUE_EVENTUALLY(condition());
```
Comment (ihsandemir) [inline hazelcast/test/src/compact/compact_schema_replication_on_cluster_restart_test.h:93]: First make sure that client is connected back and then use `ASSERT_TRUE_ALL_THE_TIME` and wait for 1 sec.

**Review by ihsandemir — CHANGES_REQUESTED** [review]
`default_schema_service::get(int64_t schemaId)` should cache the read schema to avoid more lookups.

Can you also better name the test `CompactMix`, it is not self explanatory name.

File: hazelcast/include/hazelcast/client/protocol/ClientMessage.h (hunk 6 above)
Comment (OzanCansel) [reply hazelcast/include/hazelcast/client/protocol/ClientMessage.h:934]: Agree but the problem is `field_descriptor` in Java and C++ is different, so `get<std::vector<field_descriptor>>()` is not usable.

File: hazelcast/src/hazelcast/client/compact.cpp (hunk 7 above)
Comment (OzanCansel) [reply hazelcast/src/hazelcast/client/compact.cpp:1919]: Fixed.

File: hazelcast/test/src/compact/compact_schema_replication_on_cluster_restart_test.h (hunk 8 above)
Comment (OzanCansel) [reply hazelcast/test/src/compact/compact_schema_replication_on_cluster_restart_test.h:93]: Fixed.

File: hazelcast/test/src/compact/serialization/a_type.h [hunk 9]
```diff
@@ -41,7 +41,7 @@ struct hz_serializer<test::compact::a_type> : compact::compact_serializer
     static void write(const test::compact::a_type& object,
                       compact::compact_writer& writer)
     {
-        writer.write_boolean("x", object.x);
+        writer.write_int32("x", object.x);
```
[the remark this answers is not in the document]
Comment (OzanCansel) [reply hazelcast/test/src/compact/serialization/a_type.h:44]: `a_type` is a descriptive name, in other words it tells don't bother about me, im just `a_type` :)

File: hazelcast/include/hazelcast/client/protocol/ClientMessage.h (hunk 6 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/protocol/ClientMessage.h:934]: ok

File: hazelcast/src/hazelcast/client/compact.cpp [hunk 10]
```diff
@@ -1887,12 +1887,36 @@ default_schema_service::get(int64_t schemaId)
...
+    auto invocation =
+      spi::impl::ClientInvocation::create(context_, message, SERVICE_NAME);
+    auto response = invocation->invoke().get();
+
+    response.skip_frame();
+    auto sch = response.get_nullable<schema>();
+
+    if (!sch) {
         throw exception::illegal_state{ "default_schema_service::get",
                                         "Schema doesn't exist for this type" };
```
Comment (ihsandemir) [inline hazelcast/src/hazelcast/client/compact.cpp:1916]: please print schema id.

File: hazelcast/src/hazelcast/client/network.cpp (hunk 4 above)
Comment (OzanCansel) [reply hazelcast/src/hazelcast/client/network.cpp]: Fixed.

File: hazelcast/test/src/compact/serialization/a_type.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (OzanCansel) [reply hazelcast/test/src/compact/serialization/a_type.h:44]: Fixed.

File: hazelcast/src/hazelcast/client/compact.cpp (hunk 10 above)
Comment (OzanCansel) [reply hazelcast/src/hazelcast/client/compact.cpp:1916]: Fixed.

**Review by akeles85 — DISMISSED** [review]

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**